polling_stations/apps/uk_geo_utils/helpers.py

```diff
@@ -16,6 +16,29 @@ def get_onsud_model():
     return get_model('ONSUD_MODEL', 'uk_geo_utils.Onsud')
 
 
+class Postcode:
+
+    def __init__(self, postcode):
+        self.postcode = re.sub('[^A-Z0-9]', '', str(postcode).upper())
+
+    def __str__(self):
+        return self.without_space
+
+    @property
+    def territory(self):
```

Comment:
FYI, there's a Ruby Gem that works out the territory (and does other things) here:

https://github.com/threedaymonk/uk_postcode/blob/master/lib/uk_postcode/country_lookup.rb

Might be worth taking a look at it's API for ideas.

I say this because I'm not sure 'GB' is a territory technically – I've used it in EE to mean a 'sub-national area'. Maybe it doesn't matter too much, but I was assuming that `.territory` used elsewhere in this commit would return one of ENG, SCT, WLS, NIR.

Reply:
If we want to actually work out what country (or local authority, or ward..) a postcode is in, that should be handled in the `Geocoder` class(es) rather than trying to overload everything into the `Postcode` class.

I've put this in here because there's a convenient shortcut/picture check we can do but maybe the best way is to say this check should just go back into the WhereDIV specific code ( https://github.com/DemocracyClub/UK-Polling-Stations/pull/1065/files#diff-0e31cd62c86a0e4133b1d5916504b644L258 ) and it doesn't belong in the `Postcode` class.

Trying to use a regex to work out if a postcode is in Wales (or whatever) is not a good idea for several reasons.

1. It doesn't work: For example, here are some postcodes which contain UPRNs which are in both England and Wales:
SY156AU
SY156DF
SY156DS
SY156EB

2. Even if it did work, you've got to attempt to "maintain" hundreds of characters of regex or generate it each time a new addressbase comes out or whatever. This is what the `Geocoder` classes are there to save us from.